src/ws/pam_cockpit_cert.c

```diff
@@ -0,0 +1,203 @@
+#define MAX_PEER_CERT_SIZE 100000
+
+/* Parse the module arguments */
+static void
+parse_args (int argc,
+            const char **argv)
+{
+  for (int i = 0; i < argc; i++)
+    {
+      if (strcmp (argv[i], "debug") == 0)
+        enable_debug = 1;
+      else
+        error ("invalid option: %s", argv[i]);
+    }
+}
+
+static int
+sssd_map_certificate (const char *certificate, char** username)
+{
+  int result = PAM_SERVICE_ERR;
+  sd_bus_error err = SD_BUS_ERROR_NULL;
+  sd_bus *bus = NULL;
+  sd_bus_message *user_obj_msg = NULL;
+  const char *user_obj_path = NULL;
+  int r;
+
+  r = sd_bus_open_system (&bus);
+  if (r < 0)
+    {
+      error ("Failed to connect to system bus: %s", strerror (-r));
+      result = PAM_AUTHINFO_UNAVAIL;
+      goto out;
+    }
+
+  r = sd_bus_call_method (bus,
+                          "org.freedesktop.sssd.infopipe",
+                          "/org/freedesktop/sssd/infopipe/Users",
+                          "org.freedesktop.sssd.infopipe.Users",
+                          "FindByCertificate",
+                          &err,
+                          &user_obj_msg,
+                          "s",
+                          certificate);
+
+  if (r < 0)
+    {
+      /* The error name is a bit confusing, and this is the common case; translate to readable error */
+      if (sd_bus_error_has_name (&err, "sbus.Error.NotFound"))
+        {
+          error ("No matching user for certificate");
+          result = PAM_USER_UNKNOWN;
+          goto out;
+        }
+
+      error ("Failed to map certificate to user: [%s] %s", err.name, err.message);
+      result = PAM_AUTHINFO_UNAVAIL;
+      goto out;
+    }
+
+  assert (user_obj_msg);
+
+  r = sd_bus_message_read (user_obj_msg, "o", &user_obj_path);
+  if (r < 0)
+    {
+      error ("Failed to parse response message: %s", strerror (-r));
+      goto out;
+    }
+
+  debug ("certificate mapped to user object path %s", user_obj_path);
+
+  r = sd_bus_get_property_string (bus,
+                                  "org.freedesktop.sssd.infopipe",
+                                  user_obj_path,
+                                  "org.freedesktop.sssd.infopipe.Users.User",
+                                  "name",
+                                  &err,
+                                  username);
+
+  if (r < 0)
+    {
+      error ("Failed to map user object to name: [%s] %s", err.name, err.message);
+      goto out;
+    }
+
+  assert (*username);
+  debug ("mapped certificate to user %s", *username);
+  result = PAM_SUCCESS;
+
+out:
+  sd_bus_error_free(&err);
+  sd_bus_message_unref(user_obj_msg);
+  sd_bus_unref(bus);
+  return result;
+}
+
+PAM_EXTERN int
+pam_sm_authenticate (pam_handle_t *pamh,
+                     int flags,
+                     int argc,
+                     const char **argv)
+{
+  int result = PAM_IGNORE;
+  int r;
+  const char *pam_user = NULL;
+  char cert_pem[MAX_PEER_CERT_SIZE];
```

Comment:
Are we comfortable with a stack allocation here? It seems risky from a security perspective.

Reply:
I defer to your judgement for that, I don't have enough experience with this to decide which is better. I'll change it to `mallocx()`.

Reply:
I have no qualms about security here, but I actually dislike the hardcoded arbitrary limit.